spdlog/src/sink/mod.rs

```diff
@@ -50,22 +51,94 @@ pub use std_stream_sink::*;
 pub use win_debug_sink::*;
 pub use write_sink::*;
 
-use crate::{formatter::Formatter, sync::*, ErrorHandler, Level, LevelFilter, Record, Result};
+use crate::{
+    formatter::{Formatter, FullFormatter},
+    sync::*,
+    Error, ErrorHandler, Level, LevelFilter, Record, Result,
+};
 
-/// Represents a sink
-pub trait Sink: Sync + Send {
-    /// Determines if a log message with the specified level would be logged.
+pub(crate) const SINK_DEFAULT_LEVEL_FILTER: LevelFilter = LevelFilter::All;
+
+pub(crate) type SinkErrorHandler = Atomic<Option<ErrorHandler>>;
+
+cfg_if::cfg_if! {
+    if #[cfg(test)] {
+        crate::utils::const_assert!(Atomic::<SinkErrorHandler>::is_lock_free());
+    }
+}
+
+pub struct SinkProp {
+    level_filter: Atomic<LevelFilter>,
+    formatter: RwLockMappable<Box<dyn Formatter>>,
+    error_handler: SinkErrorHandler,
+}
+
+impl Default for SinkProp {
+    fn default() -> Self {
+        Self {
+            level_filter: Atomic::new(SINK_DEFAULT_LEVEL_FILTER),
+            formatter: RwLockMappable::new(Box::new(FullFormatter::new())),
+            error_handler: Atomic::new(None),
+        }
+    }
+}
+
+impl SinkProp {
+    /// Gets the log level filter.
     #[must_use]
-    fn should_log(&self, level: Level) -> bool {
-        self.level_filter().test(level)
+    pub fn level_filter(&self) -> LevelFilter {
+        self.level_filter.load(Ordering::Relaxed)
     }
 
-    /// Logs a record.
-    fn log(&self, record: &Record) -> Result<()>;
+    /// Sets the log level filter.
+    pub fn set_level_filter(&self, level_filter: LevelFilter) {
+        self.level_filter.store(level_filter, Ordering::Relaxed)
+    }
 
-    /// Flushes any buffered records.
-    fn flush(&self) -> Result<()>;
+    /// Gets the formatter.
+    ///
+    /// TODO(doc): Note that the returned value is a lock guard.
+    pub fn formatter<'a>(&'a self) -> impl Deref<Target = dyn Formatter> + 'a {
+        RwLockMappableReadGuard::map(self.formatter.read(), |f| &**f)
+    }
```

Comment:
RPIT is used here so that we can hide what kind of lock we are using as implementation details.